ui/app/selectors/transactions.js

```diff
@@ -227,14 +231,25 @@ export const nonceSortedTransactionsSelector = createSelector(
         const nonceProps = nonceToTransactionsMap[nonce]
         insertTransactionByTime(nonceProps.transactions, transaction)
 
+        const { primaryTransaction: { time: primaryTxTime = 0 } = {} } = nonceProps
         if (status in PRIORITY_STATUS_HASH) {
-          const { primaryTransaction: { time: primaryTxTime = 0 } = {} } = nonceProps
-
           if (status === CONFIRMED_STATUS || txTime > primaryTxTime) {
             nonceProps.primaryTransaction = transaction
           }
         }
 
+        // If the primary transaction is failed or this transaction newer, assign as primary
+        if (
+          (nonceProps.primaryTransaction.status === FAILED_STATUS && nonceProps.primaryTransaction?.txReceipt?.status === '0x0' && transaction.status !== FAILED_STATUS) ||
```

Comment:
There are two failure states we'd want to catch here:

```primaryTransaction.status === FAILED``` and ```primaryTransaction.txReceipt.status === '0x0'``` these are two separate failure states. The latter could be `primaryTransaction.status === CONFIRMED```

additionally you don't want to drop the CONFIRMED check, here..so a stab at the logic might be:

```js
const previousPrimaryIsFailed = nonceProps.primaryTransaction.status === FAILED_STATUS || nonceProps.primaryTransaction?.txReceipt?.status === '0x0';

if (status === CONFIRMED_STATUS || previousPrimaryIsFailed || txTime > primaryTxTime) {
  // do stuff
}
```

Reply:
Hmm. That doesn't seem _quite_ right either. We want to treat the two types of failure (on-chain and network) in _opposite_ ways, not in the same way.

e.g. something like this:

```
const previousPrimaryIsNetworkFailure= nonceProps.primaryTransaction.status === FAILED_STATUS && nonceProps.primaryTransaction?.txReceipt?.status !== '0x0';
const currentTransactionIsOnChainFailure = transaction.txReceipt?.status === '0x0'

if (status === CONFIRMED_STATUS || currentTransactionIsOnChainFailure || previousPrimaryIsNetworkFailure || txTime > primaryTxTime) {
  // do stuff
}
```

So there are three priorities here. _Anything_ should be prioritized over a network failure, but a status of `confirmed` or an on-chain failure should be prioritized above `pending`, because they both represent a "final" state for the transaction that can't change from there on.

Reply:
Though, we may be able to assume that on-chain failures have a status of `confirmed`. I was unable to confirm that it was possible for an on-chain failure to get assigned a `failed` status.

Reply:
We're off on something because, per my PR's console statements, the primary transaction is being replaced *every time*.

Reply:
Oh! Sorry I was imagining my snippet as being _inside_ the `if (status in  PRIORITY_STATUS_HASH)` block. Though that wouldn't have made sense either because `FAILED` isn't in that object.

So maybe something like this:

```
f (status === CONFIRMED_STATUS || currentTransactionIsOnChainFailure || previousPrimaryIsNetworkFailure || (txTime > primaryTxTime && status in PRIORITY_STATUS_HASH)) {
```